### ISSUE TITLE
Allow to unset env variables

### ISSUE THREAD

**Issue by mpdude** [actions/runner#1126]
`echo 'SOME_VAR=some-value' >> $GITHUB_ENV` can be used to set an environment variable for subsequent steps, but there is no way to _unset_ it.

Note that `echo 'SOME_VAR=' >> $GITHUB_ENV` will set it to the _empty_ value, but that's different from not having the variable set.

Rationale: In some workflows, it might be necessary to checkout and/or run untrusted code. At that point I'd like to clean up as much sensitive values as possible. Unsetting env vars (like `SSH_AGENT_PID` or `SSH_AUTH_SOCK`) is part of this.

**Comment by fhammerl** [comment]
Thank you for reporting this issue. Would you mind providing an example workflow to illustrate the issue?

**Comment by mpdude** [comment]
Sure! It's not a full workflow, but I hope it gets the message across...?

```
            -   uses: aws-actions/configure-aws-credentials@v1
                with:
                    aws-access-key-id: ${{ secrets.ACCESS_KEY_ID }}
                    aws-secret-access-key: ${{ secrets.SECRET_ACCESS_KEY }}
            -   id: login-ecr
                uses: aws-actions/amazon-ecr-login@v1
            -   run: docker pull some-private/image-from-ECR
            -   run: yarn install
```

I don't trust `yarn install` specifically, especially with regard to lifecycle scripts. `--ignore-scripts` doesn't entirely cut it because sooner or later I will have to run at least some of them (details).

So, basically what I would like to do is to clean up as much as possibly after I've done the parts that I trust more, before I start running the "risky" things. 

There is not always an option of splitting such tasks across separate workflows, which would be the best solution obviously.

**Comment by mpdude** [comment]
Wait… could I just remove lines from `$GITHUB_ENV` (`sed …`) and that would do it…?

**Comment by ethomson** [comment]
This seems like a useful addition.  I'll put this on the backlog.  Thanks for the feedback.

**Comment by chrisd8088** [comment]
Another use case example for this functionality turned up while working on git-lfs/git-lfs#5236, namely, that the `ruby/setup-ruby@v1` action [sets](https://github.com/ruby/setup-ruby/blob/4b2d1d631efa087f8896c15a0c6023dc2f483198/index.js#L177) the `TMPDIR` environment variable for subsequent steps, but for the Git LFS project's CI test suite, we need to unset it before running various non-Ruby test scripts.  Otherwise, we find that `mktemp(1)` creates temporary directories under the `TMPDIR` location but Go language programs using `os.TempDir()` create them elsewhere, on another volume, and the files cannot then be moved between the two locations.

Resolving this issue while still upgrading to the current `ruby/setup-ruby@v1` action from the `actions/setup-ruby@v1` action required us to use `env -u TMPDIR` in multiple locations.  It would be much easier and more robust if we could simply unset a given environment variable by writing once to a GitHub Actions environment file, like the onoe specified by `GITHUB_ENV`, as soon as we're done with our Ruby-related actions and want to move on to our Go-related actions.

**Comment by Constantin07** [comment]
I'm facing the same issue - I need to cleanup AWS creds from the `aws-actions/configure-aws-credentials` step but don't know how to do it properly.

**Comment by michielvermeir** [comment]
> I'm facing the same issue - I need to cleanup AWS creds from the aws-actions/configure-aws-credentials step but don't know how to do it properly.

Yep, that'd also be my use case. The problem for me is that environment variables exported by `aws-actions/configure-aws-credentials` always have a higher precedence in the AWS CLI credentials chain than if I were to set `AWS_PROFILE`.

**Comment by AlexAtkinson** [comment]
For GITHUB_ENV, you can tear out content with sed. Note specifically that this is simply a file that's sourced, so you can work on it in the same way you might a .bashrc file.

```
sed -i "/badvar=/d" $GITHUB_ENV
```

I cut a gist to help explain these convenience files.
https://gist.github.com/AlexAtkinson/f9086c4a88a28cfdce5dbb1b15e947ea

**Comment by mpdude** [comment]
At least two years ago, that did not work:

https://github.com/actions/runner/issues/1126#issuecomment-853046696

**Comment by kmcentush** [comment]
Has anyone found a workaround? I also was not able to make `sed` work.

**Comment by Brikaa** [comment]
I believe the sed solution won't work with multiline values

**Comment by Moulick** [comment]
This is becoming a big issue when we want to utilize some third-party action that sets some env variables that then we cannot unset. This becomes an even bigger issue when we want to utilize that action several times and handle the output in a custom way. 

I am writing a `composite` action to invoke `aws-actions/configure-aws-credentials@v3` multiple times and stuff the creds into separate profiles. But I cannot unset the `AWS_*` variables being set by the action 😢. This leads to following steps still having at-least `AWS_SESSION_TOKEN=""` and create problems

**Comment by mossad-zika** [comment]
please help

**Comment by francardoso93** [comment]
Use SED still doesn't work, as explained in this comment: https://github.com/actions/runner/issues/1126#issuecomment-853046696

**Comment by bryanmacfarlane** [comment]
Yeah, you can't unset because each step is run out of proc and the toolkits export variable function informs the runner on variables to set on future steps: https://github.com/actions/toolkit/tree/main/packages/core#exporting-variables 

I believe it will also short circuit if the value is "" but might be worth validating.  Would probably need an unset toolkit / runner support.  This all does create odd coupling between steps though :/

**Comment by NRodrigues001** [comment]
Can someone post a link to the list where we can add customers who are asking for this feature? This is a feature Experian is very interested in. I didn't see one in the issue/discussion. Thanks!

cc @yasiralibrahem

**Comment by kellertk** [comment]
Does something like this work?

```
- uses: actions/github-script@v7
  id: unset-env-vars
  with:
  script: |
    const variables = [
      'ENV_VAR_NAME',
    ];

    for (const variable of variables) {
      if (process.env[variable]) {
        core.debug(`${variable} is set, removing`);
        delete process.env[variable];
      }
    }
```

**Comment by Moulick** [comment]
@kellertk  I'll try

**Comment by Moulick** [comment]
@kellertk nope, does not work :(

**Comment by chrisgustavsen-wf** [comment]
@bryanmacfarlane We are encountering this issue as well. For example - https://github.com/awslabs/aws-sdk-kotlin/issues/1281. Any updates on priority or workarounds?

**Comment by michaelschuett-tomtom** [comment]
not a real work around but I was able to do this which at least works for my usecase.

```
    - name: Set environment
      shell: sh
      run: |
          set -e
          cp $GITHUB_ENV /tmp/backup
          echo "SECRET=${{ inputs.secret }}" >> $GITHUB_ENV
```

then after you are done you can simply `mv /tmp/backup $GITHUB_ENV`

**Comment by runlevel5** [comment]
@kellertk yours does not work simply because you only unset the environment variables and NOT modify the `GITHUB_ENV` file so the next step would reload the environment variables from that file again.

**Comment by runlevel5** [comment]
The `sed -i "/badvar=/d" $GITHUB_ENV` suggestion from @AlexAtkinson won't work too because GitHub has disallowed any operation to do with that file. Yes, even if you `cp $GITHUB_ENV /tmp/my_temp`, GitHub would not allow you to work with that copy.

I guess it is all about security, still it is ANNOYING that GitHub does not offer a way

**Comment by jebeaudet** [comment]
Adding my voice to the problem, we have an action to fetch a secured credential using OIDC like this(workflow simplified for demo purposes) : 
```
    - name: Assume OIDC role
      uses: aws-actions/configure-aws-credentials@v4
      with:
        role-to-assume: arn:aws:iam::1234:role/some-secured-role
        aws-region: us-east-2

    - name: Get Secret value
      id: get_secret_value
      shell: bash
      env:
        AWS_REGION: us-east-2
      run: |
          export SECRET_VALUE=$(\
            aws secretsmanager get-secret-value \
            --secret-id arn:aws:secretsmanager:us-east-2:1234:secret:/some/secret \
            --query SecretString \
            --output text \
          )
          if [[ "${{ inputs.jq_escape }}" == "true" ]]; then
            SECRET_VALUE=$(echo "$SECRET_VALUE" | jq -c .)
          fi
          echo "::add-mask::$SECRET_VALUE"
          echo "secret-value=$SECRET_VALUE" >> "$GITHUB_OUTPUT"

    - name: Unassume OIDC role and environment variables cleanup
      shell: bash
      run: |
        echo "AWS_ACCESS_KEY_ID=" >> $GITHUB_ENV
        echo "AWS_SECRET_ACCESS_KEY=" >> $GITHUB_ENV
        echo "AWS_SESSION_TOKEN=" >> $GITHUB_ENV
        echo "AWS_DEFAULT_REGION=" >> $GITHUB_ENV
        echo "AWS_REGION=" >> $GITHUB_ENV
```

This works fine to clear out the creds, but we're left with `AWS_REGION` and `AWS_DEFAULT_REGION` that are set with empty value. Boto3 and the Java AWS sdk do not like this and we end up with errors like 
```
Caused by: java.lang.IllegalArgumentException: Request endpoint must have a valid hostname, but it did not: https://ssm/..amazonaws.com
```
```
File "botocore/endpoint.py", line 400, in create_endpoint
ValueError: Invalid endpoint: https://sts..amazonaws.com/
```

I know I could run this step in another job, but for simplicity and speed purposes, it would be much simpler to be able to properly _unset_ environment variables. My only workaround for this is to set a default region at the end of the action but this does not make sense at all.

**Comment by jebeaudet** [comment]
> Adding my voice to the problem, we have an action to fetch a secured credential using OIDC like this(yaml simplified for demo purposes) :
> 
> ```
>     - name: Assume OIDC role
>       uses: aws-actions/configure-aws-credentials@v4
>       with:
>         role-to-assume: arn:aws:iam::1234:role/some-secured-role
>         aws-region: us-east-2
> 
>     - name: Get Secret value
>       id: get_secret_value
>       shell: bash
>       env:
>         AWS_REGION: us-east-2
>       run: |
>           export SECRET_VALUE=$(\
>             aws secretsmanager get-secret-value \
>             --secret-id arn:aws:secretsmanager:us-east-2:1234:secret:/some/secret \
>             --query SecretString \
>             --output text \
>           )
>           if [[ "${{ inputs.jq_escape }}" == "true" ]]; then
>             SECRET_VALUE=$(echo "$SECRET_VALUE" | jq -c .)
>           fi
>           echo "::add-mask::$SECRET_VALUE"
>           echo "secret-value=$SECRET_VALUE" >> "$GITHUB_OUTPUT"
> 
>     - name: Unassume OIDC role and environment variables cleanup
>       shell: bash
>       run: |
>         echo "AWS_ACCESS_KEY_ID=" >> $GITHUB_ENV
>         echo "AWS_SECRET_ACCESS_KEY=" >> $GITHUB_ENV
>         echo "AWS_SESSION_TOKEN=" >> $GITHUB_ENV
>         echo "AWS_DEFAULT_REGION=" >> $GITHUB_ENV
>         echo "AWS_REGION=" >> $GITHUB_ENV
> ```
> 
> This works fine to clear out the creds, but we're left with `AWS_REGION` and `AWS_DEFAULT_REGION` that are set with empty value. Boto3 and the Java AWS sdk do not like this and we end up with errors like
> 
> ```
> Caused by: java.lang.IllegalArgumentException: Request endpoint must have a valid hostname, but it did not: https://ssm/..amazonaws.com
> ```
> 
> ```
> File "botocore/endpoint.py", line 400, in create_endpoint
> ValueError: Invalid endpoint: https://sts..amazonaws.com/
> ```
> 
> I know I could run this step in another job, but for simplicity and speed purposes, it would be much simpler to be able to properly _unset_ environment variables. My only workaround for this is to set a default region at the end of the action but this does not make sense at all.

**Comment by jebeaudet** [comment]
I've dug a bit in the codebase and it seems that on each step, the runner service will generate a new empty `set_env_${someGuid}` in the `/opt/actions-runner/_work/_temp/_runner_file_commands`. The step can add variable in this file and they will get processed at the end of the step by the runner service [here](https://github.com/actions/runner/blob/main/src/Runner.Worker/FileCommandManager.cs#L141) I think. This is feeded into a global map that is then passed along on each further step. 

This is why the `sed` or `cp` option does not work, the `GITHUB_ENV` file only holds new variables that will get propagated downstream. It is always empty at the beginning of each step.

Now the only way I see to do this would be to have a step before `aws-actions/configure-aws-credentials` that would use something like [inotify or other options](https://superuser.com/questions/181517/how-to-execute-a-command-whenever-a-file-changes) to watch the files in this folder and overwrite them, before the `configure-aws-credentials` step completes, but oh man that would be a solid hack.

**Comment by wonko** [comment]
As quite some people on this ticket are explicitly searching for this to solve the AWS_ env vars issues, I quickly put together a simple action which will fill a profile based of the GHA OIDC token exchange/role-assumption. It only solves my use case (GHA OIDC). It allows to add multiple profiles. 

https://github.com/krane-labs/gha-aws-oidc-profiles

**Comment by duaneking** [comment]
I'm running into this issue right now where the default install of runner software images is actively conflicting with a build pipeline; I'm under the impression that if I can just unset one single value I can unblock the build and have a working pipeline.

How do I remove this single environment value that is being set in the image by default without my consent so that I can unblock my team?

This is absolutely critical, and if we can't do this then it's a defect.

**Comment by estiller** [comment]
We are running into this issue as well. It is a sorely missing feature.

**Comment by rvesse** [comment]
Another example use case.  Aqua Security Trivy has bug https://github.com/aquasecurity/trivy-action/issues/422 because it's a `composite` action and it relies on using `$GITHUB_ENV` to set environment variables to pass into the actual `trivy` invocation.  But if you have a workflow that calls their action twice, inputs from your first call are already present in the environment and get passed into the subsequent action call overriding inputs you might be trying to change on a subsequent call

**Comment by mcandre** [comment]
The Rust setup workflow is opinionated, forcing an awkward unset behavior.

As long as environment variables (particularly for Node.js) involve unsetting then the core Actions design should provide a whole-job unset parameter that executes after every step, ensuring that the relevant variables are reliably unbound.

**Comment by duaneking** [comment]
> Thank you for reporting this issue. Would you mind providing an example workflow to illustrate the issue?

I'm currently not able to sign binaries with Github because Digicert is dumb and actively looks for the Android home value in the environment and then tries to install and fails to install things when it finds it, so I want to unset it so that I can unbreak my build and automate signing new build of our product so I can have full continuous delivery, and Github is now actively in the way of engineering excellence efforts at our company.

**Comment by bondz** [comment]
The ability to unset the env variables for a step in the workflow will be a security win.

There's a number of secrets that are defined as org or repo secrets that are available in actions. Actions can itself generate secrets now too with OIDC, (AWS, npm, etc)

It should be possible to allowlist or disallow a step from getting all the environment variables in the action - a good example is package managers that have to install external packages and may run scripts.

The recommended step for installing npm for example, should be

```yaml
steps:
  - name: Install Dependencies
    env:
      blocklist:
        vars:
          - DB_PASSWORD # explicitly remove this var
        sources:
          - step # remove all vars set via $GITHUB_ENV or outputs?
    run: npm ci

```

**Comment by mcandre** [comment]
Resetting environment variables manually (e.g. a direct `unset <var>` command) is risky. The variable is disconnected from automatic log sanitization.

Note that variables loaded by a manner other than official CI/CD secrets management, risks leaks.

**Comment by gladykov** [comment]
My use case? Unset
```
SSH_AUTH_SOCK
SSH_AGENT_PID
```

propagated by `webfactory/ssh-agent` action, to be able to use `git config --local http.https://github.com/.extraheader "AUTHORIZATION: basic` workaround (which was pain to find) to push to internal repo.

Woarkarounds on top of workarounds :/
